### PR TITLE
Ensure stale dataset references are removed

### PR DESCRIPTION
**Pull request by dstandish**
If a DAG or task previously referenced a dataset but no longer does, that reference should be removed from the DB the next time the dag is processed.

todo: check backpopulates / backrefs

### CODE REVIEW

File: airflow/migrations/versions/0114_2_4_0_add_dataset_model.py [hunk 1]
```diff
@@ -94,6 +99,12 @@ def _create_task_outlet_dataset_reference_table():
             name="todr_dataset_fkey",
             ondelete="CASCADE",
         ),
+        sa.ForeignKeyConstraint(
+            columns=('dag_id',),
+            refcolumns=['dag.dag_id'],
+            name='todr_dag_id_fkey',
```
Comment (ashb) [inline airflow/migrations/versions/0114_2_4_0_add_dataset_model.py:105]: Could you make this name match the same pattern we use on the rest of the FKs please.

File: tests/models/test_dag.py [hunk 2]
```diff
@@ -692,14 +692,14 @@ def test_bulk_write_to_db(self):
                 assert row[0] is not None
 
         # Re-sync should do fewer queries
-        with assert_queries_count(8):
+        with assert_queries_count(16):
```
Comment (ashb) [inline tests/models/test_dag.py]: This is a lot of extra queries it's now doing now. What's the cause of it?

Comment (dstandish) [reply tests/models/test_dag.py]: so, before, we only needed look at the python object to determine whether we needed to do any data operations.  if there were dataset references, we'd add them.

however we need to also _remove_ them from database if they were there before but are no longer on the python objects.

so that means even for dags and tasks that don't have any references, we have to check if they did _before_.  and for ones that _do_ have references, we have to check that there aren't any in the DB that need to be deleted.

so given that, it makes sense that queries would increase.  now, whether the current approach is _optimal_ from a performance perspective, that i'm not sure of, but i wanted to at least get it working in a reasonable way.

Comment (dstandish) [reply tests/models/test_dag.py]: i think i figured out how to reduce the queries... but now i can't seem to make the tests fail... will pick it up tomorrow

File: airflow/models/dag.py [hunk 3]
```diff
@@ -3018,7 +3039,18 @@ class DagModel(Base):
     parent_dag = relationship(
         "DagModel", remote_side=[dag_id], primaryjoin=root_dag_id == dag_id, foreign_keys=[root_dag_id]
     )
-
+    schedule_dataset_references = relationship(
```
Comment (dstandish) [inline airflow/models/dag.py:3069]: query.options where you need it

File: airflow/models/dag.py [hunk 4]
```diff
@@ -2705,24 +2709,39 @@ class InletRef(NamedTuple):
...
-        # store task-outlet-dataset references
-        for outlet_ref in outlet_references:
-            session.merge(
-                TaskOutletDatasetReference(
-                    dataset_id=stored_datasets[outlet_ref.uri].id,
-                    dag_id=outlet_ref.dag_id,
-                    task_id=outlet_ref.task_id,
+            dag_refs_to_add = {x for x in dag_refs_needed if x not in dag_refs_stored}
+            session.bulk_save_objects(dag_refs_to_add)
+            for obj in dag_refs_stored - dag_refs_needed:
```
Comment (dstandish) [inline airflow/models/dag.py:2725]: operate on the relationship attr

File: airflow/models/dag.py [hunk 5]
```diff
@@ -2705,24 +2709,39 @@ class InletRef(NamedTuple):
...
+            task_refs_stored = set(
+                session.query(TaskOutletDatasetReference)
+                .filter(
+                    TaskOutletDatasetReference.dag_id == dag_id, TaskOutletDatasetReference.task_id == task_id
                 )
+                .all()
             )
+            task_refs_to_add = {x for x in task_refs_needed if x not in task_refs_stored}
+            session.bulk_save_objects(task_refs_to_add)
+            for obj in task_refs_stored - task_refs_needed:
```
Comment (dstandish) [inline airflow/models/dag.py:2743]: possible bulk delete

File: airflow/models/dag.py [hunk 6]
```diff
@@ -2561,12 +2560,16 @@ def bulk_write_to_db(cls, dags: Collection["DAG"], session=NEW_SESSION):
...
-            session.query(DagModel)
-            .options(joinedload(DagModel.tags, innerjoin=False))
-            .filter(DagModel.dag_id.in_(dag_ids))
+            (
+                session.query(DagModel)
+                .options(joinedload(DagModel.tags, innerjoin=False))
+                .filter(DagModel.dag_id.in_(dag_ids))
+            )
+            .options(joinedload(DagModel.schedule_dataset_references))
+            .options(joinedload(DagModel.task_outlet_dataset_references))
```
Comment (uranusjr) [inline airflow/models/dag.py]:
```suggestion
            session.query(DagModel)
            .options(joinedload(DagModel.tags, innerjoin=False))
            .filter(DagModel.dag_id.in_(dag_ids))
            .options(joinedload(DagModel.schedule_dataset_references))
            .options(joinedload(DagModel.task_outlet_dataset_references))
```

Looks equivalent to me? Surprising that Black allows this.

Comment (dstandish) [reply airflow/models/dag.py]: hah yeah... not sure how that happened.

File: airflow/models/dagrun.py [hunk 7]
```diff
@@ -252,19 +252,22 @@ def refresh_from_db(self, session: Session = NEW_SESSION) -> None:
...
-            query = query.filter(cls.dag_id.in_(list(set(dag_ids))))
+            query = query.filter(cls.dag_id.in_(dag_ids))
         if only_running:
-            query = query.filter(cls.state == State.RUNNING)
+            query = query.filter(cls.state == DagRunState.RUNNING)
         else:
-            query = query.filter(cls.state.in_([State.RUNNING, State.QUEUED]))
+            query = query.filter(cls.state.in_([DagRunState.RUNNING, DagRunState.QUEUED]))
         query = query.group_by(cls.dag_id)
-        return {dag_id: count for dag_id, count in query.all()}
```
Comment (dstandish) [inline airflow/models/dagrun.py]: @uranusjr i think this was probably a mistake ... i.e. the removal of `.all()`, no?

File: airflow/models/dagrun.py [hunk 8]
```diff
@@ -252,19 +252,22 @@ def refresh_from_db(self, session: Session = NEW_SESSION) -> None:
...
+        only_running: bool = False,
+        session: Session = NEW_SESSION,
+    ) -> Dict[str, int]:
         """Get the number of active dag runs for each dag."""
         query = session.query(cls.dag_id, func.count('*'))
         if dag_ids is not None:
-            # 'set' called to avoid duplicate dag_ids, but converted back to 'list'
-            # because SQLAlchemy doesn't accept a set here.
-            query = query.filter(cls.dag_id.in_(list(set(dag_ids))))
+            query = query.filter(cls.dag_id.in_(dag_ids))
```
Comment (dstandish) [inline airflow/models/dagrun.py]: @uranusjr why this change?

Comment (dstandish) [reply airflow/models/dagrun.py]: there are other places where this method is called, and they aren't all already set-like

File: airflow/models/dagrun.py (hunk 7 above)
Comment (dstandish) [reply airflow/models/dagrun.py]: i reverted the changes to this file for now.  i don't think they are required by your other changes.

File: airflow/models/dag.py [hunk 9]
```diff
@@ -2665,29 +2665,39 @@ def bulk_write_to_db(cls, dags: Collection["DAG"], session=NEW_SESSION):
...
+                dag_references[dag.dag_id].add(dataset.uri)
                 input_datasets[DatasetModel.from_public(dataset)] = None
+            curr_outlet_references = curr_orm_dag and curr_orm_dag.task_outlet_dataset_references
             for task in dag.tasks:
-                for obj in task.outlets or []:
-                    if isinstance(obj, Dataset):
-                        outlet_references.add(OutletRef(task.dag_id, task.task_id, obj.uri))
-                        outlet_datasets[DatasetModel.from_public(obj)] = None
+                dataset_outlets = [x for x in task.outlets or [] if isinstance(x, Dataset)]
+                if not dataset_outlets:
```
Comment (jedcunningham) [inline airflow/models/dag.py:2708]: Do we need this section in addition to the task outlet reconciliation further down? What we have below might be enough on its own?

**Review by jedcunningham — APPROVED** [review]
